src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -1,235 +1,277 @@
-   :type/DateTimeWithLocalTZ "TIMESTAMP WITH TIME ZONE"
-   :type/Decimal             "DECIMAL"
-   :type/Float               "FLOAT"
-   :type/Integer             "INTEGER"
-   :type/IPAddress           "INET"
-   :type/Text                "TEXT"
-   :type/Time                "TIME"
-   :type/TimeWithTZ          "TIME WITH TIME ZONE"
-   :type/UUID                "UUID"})
+(defmulti base-type->sql-type
+  "Return a map of [[metabase.types]] type to SQL string type name. Used for casting. Looks like we're just copypasting
+  this from implementations of [[metabase.test.data.sql/field-base-type->sql-type]] so go find that stuff if you need
+  to write more implementations for this."
+  {:arglists '([driver column->value table-id]), :added "0.44.0"}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
 
 (defn- cast-values
   "Certain value types need to have their honeysql form updated to work properly during update/creation. This function
   uses honeysql casting to wrap values in the map that need to be cast with their column's type, and passes through
   types that do not need casting like integer or string."
-  [column->value table-id]
-  (let [column->field (m/index-by (comp keyword #(str/replace % "_" "-") :name)
-                                  (:fields (first (table/with-fields (db/select Table :id table-id)))))]
+  [driver column->value table-id]
+  (let [type->sql-type (base-type->sql-type driver)
+        column->field  (m/index-by (comp keyword #(str/replace % "_" "-") :name)
+                                   (:fields (first (table/with-fields (db/select Table :id table-id)))))]
     (m/map-kv-vals (fn [col value]
                      (let [{base-type :base_type :as field} (get column->field col)]
-                       (if-let [sql-type (base-type->sql-type base-type)]
+                       (if-let [sql-type (type->sql-type base-type)]
                          (hx/cast sql-type value)
                          (try
-                           (metabase.driver.sql.query-processor/->honeysql :postgres [:value value field])
+                           (sql.qp/->honeysql driver [:value value field])
                            (catch Exception e
                              (throw (ex-info (str "column cast failed: " col)
-                                             {:column col
+                                             {:column      col
                                               :original-ex (ex-message e)
                                               :status-code 400})))))))
                    column->value)))
 
+(def ^:private ^:dynamic ^Connection *connection*
+  "Data warehouse JDBC Connection to use for doing CRUD Actions. Bind this to reuse the same Connection/transaction
+  throughout a single bulk Action."
+  nil)
+
+(defn- do-with-jdbc-transaction [database-id f]
+  (if *connection*
+    (f *connection*)
+    (let [jdbc-spec (sql-jdbc.conn/db->pooled-connection-spec database-id)]
```

Comment:
I reworked the `row/*` Action impls to all use the new `with-jdbc-transaction` macro. It fixes a bunch of stuff including #23931

Reply:
> It fixes a bunch of stuff

Beyond the use of `db->pooled-connection-spec`. It would be useful to know why this is needed over/different than `jdbc/with-db-transaction` It also allows setting isolation levels and absorbs nested calls (which I believe is the reason for `*connection*` here) so a reimplementation is kinda worrying.

Reply:
The main reason is while `clojure.java.jdbc` is good for doing simple stuff you do all the time, like fetching rows from the app DB or whatever, it does a lot of additional magic that we don't always want. When we want to do complicated fancy stuff like we're doing here, IMO working around the magic is more trouble than just hitting the JDBC API directly and doing exactly what we want. 

It's not really much more code, and it's easier to reason about because I can see what's happening right there instead of trying to remember the magic weird stuff `clojure.java.jdbc` may or may not be doing.

We need custom code to implement the custom nested transaction stuff anyway and since `clojure.java.jdbc` tries to prevent them using `jdbc/with-db-transaction` made the custom nested transaction code a bit more complicated than trying to code it against stuff that was already bespoke.

We're setting the best available transaction isolation level anyway using an existing driver method, so it's not like we would have been relying on `clojure.java.jdbc` to do that for us anyway

Reply:
There are a few reasons:

1. `jdbc/with-db-transaction` "absorbs" nested transactions, but this only works if you take the transaction
connection spec bound there and pass it around explicitly to subsequent calls to `jdbc/with-db-transaction`.
This makes it hard to write bulk Actions as loops with repeated calls to single-row Actions. Of course, we could
just tweak the Actions so there was some way you could pass in an existing Connection or Connection spec for it
to use. But that would be a little more ugly and complicated than having a dynamic var. So we'd likely end up
with some sort of dynamic var and `with-` macro anyway to avoid repeated boilerplate. And it would likely make
`do-nested-transaction` harder to use or implement.

2. `jdbc/with-db-transaction` does a lot of magic that we don't necessarily want. Writing raw JDBC code is barely
any more cde and lets us have complete control over what happens and lets us see at a glance exactly what's
happening without having to keep `clojure.java.jdbc` magic in mind or work around it.